### PR TITLE
fix: js-waku / nwaku interop

### LINKED ISSUE #214: Refactor multiaddress creation
In #201 and #213 there's some code duplication as well as increased complexity in the way multiaddresses are handled when advertise addresses and/or ws/wss is used. Simplify this via the creation of a custom address factory that handles all these scenarios

### LINKED ISSUE #244: `get_waku_v2_debug_info` returns 404
Using latest master c3e42624754bc18ba50b1149c56b8c5ec3e46c9a

Start go-waku:

```
 ./waku --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --tcp-port=8001 --rpc-port=8000 --websocket-port=8002 --filter=true
```

Getting a 404 error when doing `get_waku_v2_debug_v1_info` call:

Using JS:

```
Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: false,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8003/',
    status: 404,
    statusText: 'Not Found',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} params:  [] +1s
```

Using `curl`:

```
$ curl -d '{"jsonrpc":"2.0","id":"id","method":"get_waku_v2_debug_info", "params":[]}' --header "Content-Type: application/json" http://localhost:8003               
404 page not found
```

Logs
```
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2	Version details commit=c3e4262
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2	Setting up ping protocol with duration of 20s
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2.filter	Filter protocol started
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2.relay	Relay protocol started
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2.relay	Subscribing to topic /waku/2/default-waku/proto
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2	Subscribing filter to broadcaster
2022-05-26T15:32:43.510+1000	INFO	gowaku.rpc	Rpc server started at 127.0.0.1:8000
2022-05-26T15:32:43.510+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8001/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:43.511+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8001:  enr:-Iu4QJIbHA7uVgZUL5GgriX1facdX2DQX9L_585OoFSIZ0BLYECdrRG6w5qlfqXhHDFKnjdrv985fbP8p2qAyT85i5KAgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0GFd2FrdTIF
2022-05-26T15:32:43.511+1000	INFO	gowaku.node2	Listening on /ip4/192.168.178.251/tcp/8002/ws/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:43.511+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 192.168.178.251:8002:  enr:-KO4QFyfFMJrKH-nEqb2h7AmSdi8LxAz5isEAeghW3t3OFoTLt5gJAyZxkUeUxj00tz_WVZSv7Jf4r685SX_8N5jkq2AgmlkgnY0gmlwhMCosvuKbXVsdGlhZGRyc4wACgTAqLL7Bh9C3QOJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0KFd2FrdTIF
2022-05-26T15:32:43.511+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8002/ws/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:43.512+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8002:  enr:-KO4QNwJwg5qBrkqZluctk4LfOZS4PPvuFTwvws3lXoU0g11V47_ESlCc40TrCVM3nkfruUpmClvq4Yx1YWJ_tC9w-KAgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBh9C3QOJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0KFd2FrdTIF
2022-05-26T15:32:53.523+1000	WARN	gowaku.node2	Change in host multiaddresses
2022-05-26T15:32:53.523+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8001/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:53.526+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8001:  enr:-Iu4QJIbHA7uVgZUL5GgriX1facdX2DQX9L_585OoFSIZ0BLYECdrRG6w5qlfqXhHDFKnjdrv985fbP8p2qAyT85i5KAgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0GFd2FrdTIF
2022-05-26T15:32:53.527+1000	INFO	gowaku.node2	Listening on /ip4/192.168.178.251/tcp/8002/ws/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:53.529+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 192.168.178.251:8002:  enr:-KO4QFyfFMJrKH-nEqb2h7AmSdi8LxAz5isEAeghW3t3OFoTLt5gJAyZxkUeUxj00tz_WVZSv7Jf4r685SX_8N5jkq2AgmlkgnY0gmlwhMCosvuKbXVsdGlhZGRyc4wACgTAqLL7Bh9C3QOJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0KFd2FrdTIF
2022-05-26T15:32:53.530+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8002/ws/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:53.532+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8002:  enr:-KO4QNwJwg5qBrkqZluctk4LfOZS4PPvuFTwvws3lXoU0g11V47_ESlCc40TrCVM3nkfruUpmClvq4Yx1YWJ_tC9w-KAgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBh9C3QOJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CCH0KFd2FrdTIF
2022-05-26T15:32:53.532+1000	INFO	gowaku.node2	Listening on /ip4/122.151.214.125/tcp/55750/ws/p2p/16Uiu2HAmEY3vnwibRMM3k46rp4SRUahpAPdZ3yqTmx5cjPcuxY2u
2022-05-26T15:32:53.534+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 122.151.214.125:55750:  enr:-KO4QIw-xCVCRwM5xyJ9nlJZRABHM-bGm3bjDu9DQ14mSqzYPivsWgqelacp6FRREeG4tv2bQPwYAib_6OvghsbX9NCAgmlkgnY0gmlwhHqX1n2KbXVsdGlhZGRyc4wACgR6l9Z9BtnG3QOJc2VjcDI1NmsxoQMb8Fp68-GLrWAqyY7KJ1i2Eaca9VNd55lXHINRJV28AoN0Y3CC2caFd2FrdTIF
^C2022-05-26T15:33:47.976+1000	INFO	gowaku	Received signal, shutting down...
2022-05-26T15:33:47.976+1000	ERROR	swarm2	go-libp2p-swarm@v0.10.2/swarm.go:203	error when shutting down listener: close tcp4 0.0.0.0:8002: use of closed network connection
2022-05-26T15:33:47.976+1000	INFO	gowaku.rpc	Shutting down rpc server
2022-05-26T15:33:47.976+1000	INFO	pubsub	go-libp2p-pubsub@v0.6.1/pubsub.go:626	pubsub processloop shutting down

status-im/go-waku/build  master ✔                                 6d8h  
▶ ./waku --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --log-level=debug --tcp-port=8004 --rpc-port=8003 --websocket-port=8005 --filter=true
2022-05-26T15:33:51.738+1000	INFO	gowaku.node2	Version details commit=c3e4262
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2	Setting up ping protocol with duration of 20s
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2.filter	Filter protocol started
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2.relay	Relay protocol started
2022-05-26T15:33:51.739+1000	DEBUG	pubsub	go-libp2p-pubsub@v0.6.1/gossipsub.go:1034	JOIN /waku/2/default-waku/proto
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2.relay	Subscribing to topic /waku/2/default-waku/proto
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2	Subscribing filter to broadcaster
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8004/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:33:51.739+1000	INFO	gowaku.rpc	Rpc server started at 127.0.0.1:8003
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8004:  enr:-Iu4QMAo6QzcdUlKnN0F1wR0vV6mH3mO5sdpj4DTurQZmx4tbKn2ziN8AVgnxZuMuBlE7nkgSH6_oEDJyG5ZpXE8zJCAgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0SFd2FrdTIF
2022-05-26T15:33:51.739+1000	INFO	gowaku.node2	Listening on /ip4/192.168.178.251/tcp/8005/ws/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:33:51.740+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 192.168.178.251:8005:  enr:-KO4QMe_vLtfIerhy9K1_IE3HiEVsFobC9Ca6dZSPriBxjuGJ5vzIBv5qWXMe8a-02X46Igb1SOh6XV7DGmASslIhFqAgmlkgnY0gmlwhMCosvuKbXVsdGlhZGRyc4wACgTAqLL7Bh9F3QOJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0WFd2FrdTIF
2022-05-26T15:33:51.740+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8005/ws/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:33:51.740+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8005:  enr:-KO4QP18Z9rAHdAaUTV6uVHzH_-D-r-DjiWQ66gV4r-f3SJPYWGIzL6fpCoXGSLG4vr0bzCqjKDSIsPHjPjMZepmZvKAgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBh9F3QOJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0WFd2FrdTIF
2022-05-26T15:33:56.739+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:40	DiscoverGateway address:192.168.178.1
2022-05-26T15:33:56.739+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:33:56.792+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
2022-05-26T15:34:01.745+1000	WARN	gowaku.node2	Change in host multiaddresses
2022-05-26T15:34:01.745+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8004/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8004:  enr:-Iu4QMAo6QzcdUlKnN0F1wR0vV6mH3mO5sdpj4DTurQZmx4tbKn2ziN8AVgnxZuMuBlE7nkgSH6_oEDJyG5ZpXE8zJCAgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0SFd2FrdTIF
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	Listening on /ip4/192.168.178.251/tcp/8005/ws/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 192.168.178.251:8005:  enr:-KO4QMe_vLtfIerhy9K1_IE3HiEVsFobC9Ca6dZSPriBxjuGJ5vzIBv5qWXMe8a-02X46Igb1SOh6XV7DGmASslIhFqAgmlkgnY0gmlwhMCosvuKbXVsdGlhZGRyc4wACgTAqLL7Bh9F3QOJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0WFd2FrdTIF
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	Listening on /ip4/127.0.0.1/tcp/8005/ws/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 127.0.0.1:8005:  enr:-KO4QP18Z9rAHdAaUTV6uVHzH_-D-r-DjiWQ66gV4r-f3SJPYWGIzL6fpCoXGSLG4vr0bzCqjKDSIsPHjPjMZepmZvKAgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBh9F3QOJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCH0WFd2FrdTIF
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	Listening on /ip4/122.151.214.125/tcp/18523/ws/p2p/16Uiu2HAmFcYCmuwdGd8MS8kteBreCHk9rspeNEgeLWkkkYk9LzcH
2022-05-26T15:34:01.746+1000	INFO	gowaku.node2	DNS: discoverable ENR for IP 122.151.214.125:18523:  enr:-KO4QGYPXpipEEA0Sxf3-1EER9bvBo4e2OXTP6yN2cPLD54Qb6rFWUQkrng4sjgV0cEg-rGRGOr-ANDqKVeUPqIOweyAgmlkgnY0gmlwhHqX1n2KbXVsdGlhZGRyc4wACgR6l9Z9Bkhb3QOJc2VjcDI1NmsxoQMr8lW6TdBQYwR0q39Rjwu3Jyz5yhLNLggEfP9nDqa6moN0Y3CCSFuFd2FrdTIF
2022-05-26T15:34:16.739+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:34:16.748+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
2022-05-26T15:34:36.740+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:34:36.752+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
2022-05-26T15:34:56.740+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:34:56.750+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
2022-05-26T15:35:16.739+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:35:16.747+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
2022-05-26T15:35:36.740+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:164	Attempting port map: tcp/8005
2022-05-26T15:35:36.803+1000	DEBUG	nat	go-libp2p-nat@v0.1.0/nat.go:185	NAT Mapping: 18523 --> 8005 (tcp)
^C2022-05-26T15:35:49.265+1000	INFO	gowaku	Received signal, shutting down...
2022-05-26T15:35:49.265+1000	INFO	gowaku.rpc	Shutting down rpc server
2022-05-26T15:35:49.265+1000	ERROR	swarm2	go-libp2p-swarm@v0.10.2/swarm.go:203	error when shutting down listener: close tcp4 0.0.0.0:8005: use of closed network connection
2022-05-26T15:35:49.265+1000	INFO	pubsub	go-libp2p-pubsub@v0.6.1/pubsub.go:626	pubsub processloop shutting down
```

### LINKED ISSUE #247: JSON RPC/arguments: Discrepancies with nwaku
Running the js-waku test suite with go-waku, here are some errors encountered:

Log format:

```
waku:nwaku RPC Query:  <method> <params>
waku:nwaku RPC Response:  <response headers> <response content>
```

### `get_waku_v2_relay_v1_messages`

```
 waku:nwaku RPC Query:  get_waku_v2_relay_v1_messages [ '/waku/2/default-waku/proto' ] +1s
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8003/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} {
  result: null,
  error: 'json: cannot unmarshal string into Go value of type rpc.TopicArgs',
  id: 1
}
```

### `get_waku_v2_debug_v1_info`

`enrUri` is missing

### `post_waku_v2_private_v1_asymmetric_message`

```
  waku:nwaku RPC Query:  post_waku_v2_private_v1_asymmetric_message [
  '/waku/2/default-waku/proto',
  {
    contentTopic: '/test/1/waku-message/utf8',
    payload: '4865726520697320616e20656e63727970746564206d6573736167652e'
  },
  '0x04e944b7578272f771b121410e92be5add484b66bc620068e5e208fa54291379d2a53472c5f286da08982145d6514b1a17e144b1fa4962b9c050d76303735f591e'
] +556ms
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8000/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} {
  result: null,
  error: 'json: cannot unmarshal string into Go value of type rpc.AsymmetricMessageArgs',
  id: 1
} +1ms
```

### `post_waku_v2_private_v1_symmetric_message`

```
waku:nwaku RPC Query:  post_waku_v2_private_v1_symmetric_message [
  '/waku/2/default-waku/proto',
  {
    contentTopic: '/test/1/waku-message/utf8',
    payload: '486572652069732061206d65737361676520656e6372797074656420696e20612073796d6d6574726963206d616e6e65722e'
  },
  '0xae0bbbe7a7eef7a4d95c6610824a9402c1f9a2a47f0f1daac72a4edcb7aede20'
] +543ms
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8000/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} {
  result: null,
  error: 'json: cannot unmarshal string into Go value of type rpc.SymmetricMessageArgs',
  id: 1
} +2ms
```

### `post_waku_v2_relay_v1_message`

```
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '4865726520697320616e6f74686572206d6573736167652e',
    contentTopic: '/test/1/waku-relay/utf8',
    timestamp: 1653652735961000000
  }
] +236ms
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8000/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} {
  result: null,
  error: 'json: cannot unmarshal string into Go value of type rpc.RelayMessageArgs',
  id: 1
} +2ms
```

### `--persist-messages`

```
Incorrect Usage. flag provided but not defined: -persist-messages
```

### PR DESCRIPTION
**Pull request by richard-ramos**
Fixes #244 
Fixes #247 
Partially fixes #214

### CODE REVIEW

**Comment by D4nte** [comment]
I tested 01f220ed1ff742c716430eb9dbb8928233aa94cf with js-waku:
I am getting

```
   Error: protocol selection failed
        at select (/home/fryorcraken/src/status-im/js-waku/node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (/home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
AggregateError: 
    Error: protocol selection failed
        at select (node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
    at maybeSettle (node_modules/p-some/index.js:31:11)
    at /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:69:23
```

I am dialing with the following protocols:

```
  "/vac/waku/relay/2.0.0-beta2",
  "/vac/waku/relay/2.0.0",
```

What protocols are supported? I'd expect at least one to be selected.  

I am also getting some socket hang up issues:

```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --log-level=debug --tcp-port=8001 --rpc-port=8000 --websocket-port=8002 --store=false --filter=false --lightpush=false +0ms
  waku:nwaku nwaku 359751 started at 3:11:07 pm +32ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +0ms
  waku:nwaku nwaku node has been started +35ms
  waku:nwaku RPC Query:  get_waku_v2_debug_v1_info [] +0ms
  waku:nwaku nwaku 359751 getting SIGINT at 3:11:07 pm +6ms
  waku:nwaku nwaku 359751 interrupted: true +0ms

FetchError: request to http://localhost:8000/ failed, reason: socket hang up
    at ClientRequest.<anonymous> (node_modules/node-fetch/lib/index.js:1461:11)
    at ClientRequest.emit (node:events:520:28)
    at ClientRequest.emit (node:domain:475:12)
    at Socket.socketOnEnd (node:_http_client:466:9)
    at Socket.emit (node:events:532:35)
    at Socket.emit (node:domain:475:12)
    at endReadableNT (node:internal/streams/readable:1346:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

Port 8000 is rpc as you can see for the parameters.

Also getting other errors:

```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --relay=true --rpc=true --rpc-admin=true --websocket-support=true --log-level=debug --tcp-port=8004 --rpc-port=8003 --websocket-port=8005 --persist-messages=true +7ms
  waku:nwaku nwaku 360231 started at 3:11:24 pm +12ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +0ms
  waku:nwaku nwaku 360215 process exited with 0 at 3:11:24 pm +1ms
  waku:nwaku nwaku node has been started +37ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '4d6573736167652030',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1654146684088000000
  }
] +0ms
  waku:nwaku RPC Response:  Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 4,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: true,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:8003/',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
} { result: null, error: 'illegal base64 data at input byte 16', id: 1 } +3ms
  waku:nwaku nwaku 360231 getting SIGINT at 3:11:24 pm +2ms
  waku:nwaku nwaku 360231 interrupted: true +0ms

AssertionError: expected null to be true
Expected :true
Actual   :[null]
<Click to see difference>

    at Context.<anonymous> (src/lib/waku_store/index.node.spec.ts:46:14)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

**Comment by D4nte** [comment]
Feel free to test yourself:
1. checkout the js-waku `go-waku-drop-in branch`
2. Set some env variable:
```
export DEBUG=waku* # To get the JSON query/response for debugging
export WAKU_SERVICE_NODE_BIN=/home/fryorcraken/src/status-im/go-waku/build/waku # Point the go-waku bin
export WAKU_SERVICE_NODE_DIR=/home/fryorcraken/src/status-im/go-waku/build # May not be needed, folder where the go-waku bin will be executed
```
3. run the tests: `npm run test:unit`

**Comment by richard-ramos** [comment]
```
Error: protocol selection failed
        at select (node_modules/multistream-select/src/select.js:72:17)
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
        at async Upgrader._multiplexOutbound (node_modules/libp2p/src/upgrader.js:459:36)
        at async Upgrader.upgradeOutbound (node_modules/libp2p/src/upgrader.js:204:44)
        at async ClassIsWrapper.dial (node_modules/libp2p-websockets/src/index.js:54:18)
        at async TransportManager.dial (node_modules/libp2p/src/transport-manager.js:111:14)
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/libp2p/src/dialer/dial-request.js:81:18
        at async /home/fryorcraken/src/status-im/js-waku/node_modules/p-some/index.js:53:19
 ```
This error happened because go-libp2p v0.19.0 removed mplex from the list of default muxers (which is what nim-libp2p and js-libp2p support at the moment). I added it back in `master` but forgot to rebase this branch. See https://github.com/libp2p/specs/pull/402 and https://github.com/libp2p/go-libp2p/pull/1344

---

```
FetchError: request to http://localhost:8000/ failed, reason: socket hang up
```
Fixed! 

---

We're getting better results now :)
```
  107 passing (1m)
  1 pending
  24 failing
```

I'll have a look at the `illegal base64 data at input byte 16` error, but there are some others that I need some help understanding them: https://gist.github.com/richard-ramos/147ac70ff0ec80dcfaf101eb2a198db6

File: waku.go [hunk 1]
```diff
@@ -111,6 +111,16 @@ func main() {
...
 				Destination: &options.UseDB,
 			},
+			&cli.BoolFlag{
+				Name:        "persist-messages",
+				Usage:       "Enable message persistence",
+				Destination: &options.Store.PersistMessages,
+			},
+			&cli.StringFlag{
+				Name:        "nat",
+				Usage:       "TODO",
```
Comment (alaibe) [inline waku.go:121]: Maybe add a better description even if todo?

**Review by alaibe — APPROVED** [review]
code LGTM

**Pull request merged**